File: apps/gui/src/lib/stores/profiles.ts
import { writable, derived } from 'svelte/store';

export interface GameWindowConfig {
	title: string;
	className: string;
	processName: string;
	exactMatch: boolean;
	fullscreen: boolean;
}

export interface ColorConfig {
	name: string;
	r: number;
	g: number;
	b: number;
	tolerance: number;
}

export interface ImageConfig {
	name: string;
	path: string;
	threshold: number;
	preload: boolean;
}

export interface GameTriggerConfig {
	name: string;
	type: string;
	action: string;
	target: string;
	interval: number;
	enabled: boolean;
}

export interface GameScriptConfig {
	name: string;
	path: string;
	autoStart: boolean;
	restartOnCrash: boolean;
	priority: number;
}

export interface HotkeyConfig {
	start: string[];
	stop: string[];
	pause: string[];
	emergencyStop: string[];
}

export interface GameProfile {
	id: string;
	name: string;
	version: string;
	description: string;
	window: GameWindowConfig;
	colors: ColorConfig[];
	images: ImageConfig[];
	triggers: GameTriggerConfig[];
	scripts: GameScriptConfig[];
	hotkeys: HotkeyConfig;
	settings: Record<string, string>;
}

function createProfilesStore() {
	const store = writable<GameProfile[]>([]);
	const activeId = writable<string>('');
	const invoke = (window as any).__TAURI_INVOKE__;
	let currentProfiles: GameProfile[] = [];
	let currentActiveId = '';

	store.subscribe(value => {
		currentProfiles = value;
	});
	activeId.subscribe(value => {
		currentActiveId = value;
	});

	const withDefaultHotkeys = (profile: any): GameProfile => ({
		...profile,
		hotkeys: {
			start: profile?.hotkeys?.start ?? ['F5'],
			stop: profile?.hotkeys?.stop ?? ['F6'],
			pause: profile?.hotkeys?.pause ?? ['F7'],
			emergencyStop: profile?.hotkeys?.emergencyStop ?? ['F12'],
		},
	});

	return {
		subscribe: store.subscribe,
		activeId,
		async load() {
			if (!invoke) return;
			try {
				const result = await invoke('get_profiles');
				store.set(((result || []) as any[]).map(withDefaultHotkeys));
			} catch { /* ignore */ }
		},
		async loadActive() {
			if (!invoke) return;
			try {
				const result = await invoke('get_active_profile');
				if (result?.id) activeId.set(result.id);
			} catch { /* ignore */ }
		},
		async setActive(id: string) {
			if (!invoke) {
				activeId.set(id);
				return;
			}
			try {
				await invoke('set_active_profile', { id });
				activeId.set(id);
			} catch { /* ignore */ }
		},
		async create(name: string) {
			if (!invoke) {
				const id = `profile_${Date.now()}`;
				const profile = withDefaultHotkeys({
					id,
					name,
					version: '1.0',
					description: '',
					window: { title: '', className: '', processName: '', exactMatch: false, fullscreen: false },
					colors: [],
					images: [],
					triggers: [],
					scripts: [],
					settings: {},
				});
				store.update(items => [...items, profile]);
				if (!currentActiveId) activeId.set(id);
				return id;
			}
			try {
				const id = await invoke('create_profile', { name });
				await this.load();
				return id;
			} catch { return ''; }
		},
		async remove(id: string) {
			if (!invoke) {
				store.update(items => items.filter(profile => profile.id !== id));
				if (currentActiveId === id) {
					activeId.set(currentProfiles.find(profile => profile.id !== id)?.id || '');
				}
				return;
			}
			try {
				await invoke('delete_profile', { id });
				await this.load();
			} catch { /* ignore */ }
		},
		async update(profile: GameProfile) {
			if (!invoke) {
				const normalized = withDefaultHotkeys(profile);
				store.update(items => items.map(item => item.id === normalized.id ? normalized : item));
				return;
			}
			try {
				await invoke('update_profile', { profile });
				await this.load();
			} catch { /* ignore */ }
		},
		async exportToJson(id: string) {
			if (!invoke) {
				const profile = currentProfiles.find(item => item.id === id);
				return profile ? JSON.stringify(profile, null, 2) : '';
			}
			try {
				return await invoke('export_profile_json', { id });
			} catch { return ''; }
		},
		async importFromJson(json: string) {
			if (!invoke) {
				try {
					const profile = withDefaultHotkeys(JSON.parse(json));
					store.update(items => {
						const exists = items.some(item => item.id === profile.id);
						return exists
							? items.map(item => item.id === profile.id ? profile : item)
							: [...items, profile];
					});
					return true;
				} catch {
					return false;
				}
			}
			try {
				await invoke('import_profile_json', { json });
				await this.load();
				return true;
			} catch { return false; }
		},
		loadDevData() {
			store.set([
				{
					id: 'default',
					name: '默认配置',
					version: '1.0',
					description: '默认游戏配置',
					window: { title: '', className: '', processName: '', exactMatch: false, fullscreen: false },
					colors: [],
					images: [],
					triggers: [],
					scripts: [],
					hotkeys: {
						start: ['F5'],
						stop: ['F6'],
						pause: ['F7'],
						emergencyStop: ['F12'],
					},
					settings: {},
				},
				{
					id: 'game1',
					name: '游戏配置 1',
					version: '1.0',
					description: '示例游戏配置',
					window: { title: 'MyGame', className: '', processName: 'mygame.exe', exactMatch: false, fullscreen: true },
					colors: [{ name: 'HP', r: 255, g: 0, b: 0, tolerance: 10 }],
					images: [],
					triggers: [],
					scripts: [],
					hotkeys: {
						start: ['F5'],
						stop: ['F6'],
						pause: ['F7'],
						emergencyStop: ['F12'],
					},
					settings: {},
				},
			]);
			activeId.set('default');
		},
	};
}

export const profiles = createProfilesStore();

export const activeProfile = derived(
	[profiles, profiles.activeId],
	([$profiles, $activeId]) => $profiles.find(p => p.id === $activeId) || null
);

File: apps/gui/src-tauri/src/commands/profiles.rs
use crate::hotkeys;
use serde_json::{json, Value};
use std::env;
use std::fs;
use std::path::PathBuf;
use std::time::{SystemTime, UNIX_EPOCH};
use tauri::AppHandle;

fn profiles_dir() -> PathBuf {
    let mut path = local_data_dir();
    path.push("wingman");
    path.push("profiles");
    fs::create_dir_all(&path).ok();
    path
}

fn active_profile_path() -> PathBuf {
    let mut path = profiles_dir();
    path.pop();
    path.push("active_profile.txt");
    path
}

fn profile_path(id: &str) -> PathBuf {
    // Validate that id is a safe filename to prevent path traversal attacks
    // Reject unsafe IDs instead of silently sanitizing (fail-closed approach)
    if !is_safe_filename(id) {
        panic!("Unsafe profile id detected: '{}'. Use set_active_profile() which validates before calling this.", id);
    }

    let mut path = profiles_dir();
    path.push(format!("{}.json", id));
    path
}

/// Check if a string is a safe filename (no path traversal)
fn is_safe_filename(id: &str) -> bool {
    if id.is_empty() {
        return false;
    }

    // Reject path separators and parent directory references
    if id.contains('/') || id.contains('\\') || id.contains("..") {
        return false;
    }

    // Only allow alphanumeric, underscore, hyphen, and dot
    id.chars().all(|c| c.is_alphanumeric() || c == '_' || c == '-' || c == '.')
}

/// Sanitize a filename by removing unsafe characters
fn local_data_dir() -> PathBuf {
    #[cfg(windows)]
    {
        if let Some(value) = env::var_os("LOCALAPPDATA") {
            return PathBuf::from(value);
        }
        if let Some(value) = env::var_os("APPDATA") {
            return PathBuf::from(value);
        }
    }

    #[cfg(target_os = "macos")]
    {
        if let Some(value) = env::var_os("HOME") {
            return PathBuf::from(value).join("Library").join("Application Support");
        }
    }

    #[cfg(all(unix, not(target_os = "macos")))]
    {
        if let Some(value) = env::var_os("XDG_DATA_HOME") {
            return PathBuf::from(value);
        }
        if let Some(value) = env::var_os("HOME") {
            return PathBuf::from(value).join(".local").join("share");
        }
    }

    PathBuf::from(".")
}

fn timestamp_millis() -> u128 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|duration| duration.as_millis())
        .unwrap_or_default()
}

fn load_all_profiles() -> Vec<Value> {
    let dir = profiles_dir();
    let mut profiles = Vec::new();
    if let Ok(entries) = fs::read_dir(&dir) {
        for entry in entries.flatten() {
            let path = entry.path();
            if path.extension().map_or(false, |e| e == "json") {
                if let Ok(content) = fs::read_to_string(&path) {
                    if let Ok(profile) = serde_json::from_str::<Value>(&content) {
                        profiles.push(profile);
                    }
                }
            }
        }
    }
    profiles
}

#[tauri::command]
pub async fn get_profiles() -> Result<Vec<Value>, String> {
    Ok(load_all_profiles())
}

#[tauri::command]
pub async fn get_active_profile() -> Result<Value, String> {
    let path = active_profile_path();
    if let Ok(active_id) = fs::read_to_string(&path) {
        let active_id = active_id.trim();
        if !active_id.is_empty() {
            let profile_path = profile_path(active_id);
            if let Ok(content) = fs::read_to_string(&profile_path) {
                if let Ok(profile) = serde_json::from_str::<Value>(&content) {
                    return Ok(profile);
                }
            }
        }
    }
    Ok(json!(null))
}

#[tauri::command]
pub async fn set_active_profile(app: AppHandle, id: String) -> Result<(), String> {
    // Validate id to prevent injection attacks
    if !is_safe_filename(&id) {
        return Err("Invalid profile id: contains unsafe characters".to_string());
    }

    let path = active_profile_path();
    fs::write(&path, &id).map_err(|e| e.to_string())?;
    hotkeys::reload_hotkeys(&app)?;
    Ok(())
}

#[tauri::command]
pub async fn create_profile(name: String) -> Result<String, String> {
    let id = format!("profile_{}", timestamp_millis());
    let profile = json!({
        "id": id,
        "name": name,
        "version": "1.0",
        "description": "",
        "window": { "title": "", "className": "", "processName": "", "exactMatch": false, "fullscreen": false },
        "colors": [],
        "images": [],
        "triggers": [],
        "scripts": [],
        "hotkeys": {
            "start": ["F5"],
            "stop": ["F6"],
            "pause": ["F7"],
            "emergencyStop": ["F12"]
        },
        "settings": {}
    });
    let path = profile_path(&id);
    let content = serde_json::to_string_pretty(&profile).map_err(|e| e.to_string())?;
    fs::write(&path, content).map_err(|e| e.to_string())?;
    Ok(id)
}

#[tauri::command]
pub async fn delete_profile(app: AppHandle, id: String) -> Result<(), String> {
    let path = profile_path(&id);
    if path.exists() {
        fs::remove_file(&path).map_err(|e| e.to_string())?;
    }
    // If this was the active profile, clear it
    let active_path = active_profile_path();
    if let Ok(active_id) = fs::read_to_string(&active_path) {
        if active_id.trim() == id {
            fs::write(&active_path, "").ok();
            hotkeys::reload_hotkeys(&app)?;
        }
    }
    Ok(())
}

#[tauri::command]
pub async fn update_profile(app: AppHandle, profile: Value) -> Result<(), String> {
    let id = profile["id"]
        .as_str()
        .ok_or("Profile missing id field")?
        .to_string();
    let path = profile_path(&id);
    let content = serde_json::to_string_pretty(&profile).map_err(|e| e.to_string())?;
    fs::write(&path, content).map_err(|e| e.to_string())?;

    if let Ok(active_id) = fs::read_to_string(active_profile_path()) {
        if active_id.trim() == id {
            hotkeys::reload_hotkeys(&app)?;
        }
    }

    Ok(())
}

#[tauri::command]
pub async fn export_profile_json(id: String) -> Result<String, String> {
    let path = profile_path(&id);
    fs::read_to_string(&path).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn import_profile_json(json: String) -> Result<(), String> {
    let profile: Value = serde_json::from_str(&json).map_err(|e| e.to_string())?;
    let id = profile["id"]
        .as_str()
        .ok_or("Profile missing id field")?
        .to_string();
    let path = profile_path(&id);
    let content = serde_json::to_string_pretty(&profile).map_err(|e| e.to_string())?;
    fs::write(&path, content).map_err(|e| e.to_string())?;
    Ok(())
}
